test/test_vmap.py

```diff
@@ -0,0 +1,113 @@
+from torch.testing._internal.common_utils import TestCase, run_tests
+import torch
+from torch import vmap
+
+class TestVmap(TestCase):
+    def test_non_tensor_output_raises(self):
+        with self.assertRaisesRegex(ValueError, "got type <class 'float'> as the return"):
+            output = vmap(lambda x: 3.14)(torch.ones(3))
+
+        def multiple_outputs(x):
+            return x, 3
+
+        with self.assertRaisesRegex(ValueError, "got type <class 'int'> for return 1"):
+            vmap(multiple_outputs)(torch.ones(3))
+
+    def test_different_map_dim_size_raises(self):
+        x = torch.randn(2)
+        y = torch.randn(3)
+        expected_msg = 'Expected all tensors to have the same size in the mapped dimension'
+        with self.assertRaisesRegex(ValueError, expected_msg):
+            vmap(torch.mul)(x, y)
+
+    def test_func_with_no_inputs(self):
+        expected_msg = 'got no inputs'
+
+        def foo():
+            return torch.randn(3)
+
+        def bar(x):
+            return torch.randn(3)
+
+        with self.assertRaisesRegex(ValueError, expected_msg):
+            vmap(foo)()
+
+        with self.assertRaisesRegex(ValueError, expected_msg):
+            vmap(bar)()
+
+    def test_constant_function(self):
+        output = vmap(lambda x: torch.tensor(3.14))(torch.ones(3))
+        self.assertEqual(output, torch.tensor([3.14, 3.14, 3.14]))
+
+    def test_single_input(self):
+        x = torch.randn(2, 3)
+
+        def square(x):
+            return x * x
+
+        output = vmap(square)(x)
+        self.assertEqual(output, x * x)
+
+    def test_multiple_inputs(self):
+        x = torch.randn(2, 3)
+        y = torch.randn(2, 3)
+        output = vmap(torch.mul)(x, y)
+        self.assertEqual(output, x * y)
+
+    def test_multiple_outputs(self):
+        def foo(x):
+            return x * x, x * x * x
+
+        x = torch.randn(3)
+        outputs = vmap(foo)(x)
+        self.assertEqual(outputs[0], x * x)
+        self.assertEqual(outputs[1], x * x * x)
+
+    def test_nested_with_same_map_dim(self):
+        x = torch.randn(2, 3, 5)
+        y = torch.randn(2, 3, 5)
+        output = vmap(vmap(torch.mul))(x, y)
+        self.assertEqual(output, x * y)
+
+        output = vmap(vmap(vmap(torch.mul)))(x, y)
+        self.assertEqual(output, x * y)
+
+    def test_nested_with_different_map_dim(self):
+        x = torch.randn(2, 3)
+        y = torch.randn(5, 3)
+        output = vmap(lambda x: vmap(lambda y: x * y)(y))(x)
+        self.assertEqual(output, x.view(2, 1, 3) * y)
```

Comment:
I did a double take here because I'm expecting the size to be `(2, 5, 3)`

Reply:
Yeah, I can see that. I'll add an explicit check for the size so that it is obvious that the result should be 2, 5, 3